blog/2025-08-22-pattern.mdx

```diff
@@ -0,0 +1,312 @@
+several key areas of Compact programming:
+
+- Parameters of circuits, anonymous circuits, and a constructor can be a pattern.
+- `const` binding statements can bind a pattern. This is arguably where you'll use destructuring
+most frequently for local variable assignments, allowing you to unpack values
+from expressions into new variables.
+
+## How to Destructure Values in Compact 🛠️
+
+Let's dive into the specifics of how to use tuple and structure patterns.
+The grammar for a pattern `p` is provided in
+[Compact's reference](https://docs.midnight.network/develop/reference/compact/lang-ref#parameters-patterns-and-destructuring).
+
+
+### Destructuring Tuple Values
+
+Tuple values can be destructured using a list of patterns enclosed in `[` and `]`.
+You can destructure a tuple into a sequence of identifiers:
+
+```compact
+const [a, b] = [1, true]; // 'a' is 1, 'b' is true
+```
+
+Sometimes you don't need all the values in a tuple. Compact allows you to skip
+elements by simply omitting the identifier but you cannot skip the actual element:
+
+```compact
+const [x, , z] = [10, 20, 30]; // 'x' is 10, 'z' is 30, 20 is skipped
+// const [x, , z] = [10, , 30]; // triggers a parse error
+```
+
+Skipped identifiers count towards the length of the pattern if there exists a
+named identifier after them. For example, `[x, y, ,]` has length 2 but
+`[x, , y]` has length 3.
+
+:::note
+If you want to enforce length 'l' of a tuple pattern you must have a named identifier
+at the `l-1` index of your pattern. For example, if you need to enforce a tuple
+pattern of legth 3 you need to have a named identifier on index 2 (e.g., all
+`[, , x]`, `[, x, y]`, and `[x, y, z]` enforce a tuple pattern of length 3).
+:::
+
+### Type Annotations for Tuple Patterns
+
+In a circuit definition, type declarations are required for tuple patterns:
+
+```compact
+circuit exampleTuplePattern([x, y]: [Field, Boolean]): Field {
+  // x has type Field, y has type Boolean
+  return x;
+}
+```
+
+Such a type declaration dictates the structure and the type of each element.  For example,
+in `exampleTuplePattern` the input has to be a tuple, it first element has to be a value
+of type `Field`, and its second element has to be a value of type `Boolean`.
+A call to `exampleTuplePattern` that does not meet these restrictions
+(e.g., `exampleTuplePattern([1, 1])`) will trigger a type error.
+
+For const bindings and anonymous circuits, type annotations are optional but recommended
+for clarity and an easier debugging experience. When a type annotation is provided
+ensure its structure aligns with the pattern. For example, `[a, , b] : [Field, Field, Field]`
+aligns the structure of the pattern and its type whereas `[a, , b] : [Field, Field]` does not:
+the pattern states that it is a tuple with three elements but the type annotation is
+tuple type of length two.
+
+Unlike patterns, types do not allow skipping elements in a tuple type. For example,
```

Comment:
This is an interesting one.  What is the motivation?  To me, I feel like I ought to be able to do that (omit a type for an unnamed element, I certainly can't refer to it so why do I care what the type is?  If that type changes on the right-hand side, I really don't want to have to update this type annotation for the value I don't care about.)

In fact, I'd actually like to be able to have type inference for some elements and write:

```
const [x, y, z]: [Field, , Uint<32>] = foo();
```
Where the type annotations are checked and the missing one is inferred from the return type of `foo`.

Reply:
I feel like restrictions like this are important to justify a little bit from a design standpoint.  Is there a technical reason we need it?  Do we just like it that way?  Is it something that we might change?